### ISSUE TITLE
pkcs11-tool and FAPI not provisioned and ErrorCode (0x00060034) Entities_List  errors

### ISSUE THREAD

**Issue by Dvergatal** [tpm2-software/tpm2-pkcs11#655]
Hi all,
I have sucesfully compiled tpm2-pkcs11 and it ss working for me with libp11 engine library together with openssl, but the problem is that i'm getting this errors:

```bash
ERROR:fapi:src/tss2-fapi/api/Fapi_List.c:221:Fapi_List_Finish() FAPI not provisioned. 
ERROR:fapi:src/tss2-fapi/api/Fapi_List.c:81:Fapi_List() ErrorCode (0x00060034) Entities_List 
ERROR: Listing FAPI token objects failed.
```

whenever i do something on the tpm token. Nevertheless all crypto materials are being made, the token is also initialized and all is working but i do not know why this error even occurs.

Could someone explain that to me what does it mean?

**Comment by williamcroberts** [comment]
TL;DR Its superfluous, ignore it.

The tpm2-tss project ships with library called FAPI. This has been integrated into the project so keys stored using FAPI will be discovered and integrated into the token list. However, when FAPI isn't set up you get this error, but the library internally ignores it. The logging is an artifact of how tpm2-tss project does its logging.

You can suppress the logs, aaccording to https://github.com/tpm2-software/tpm2-tss/blob/aa7595d01612eef6fd19af3545c280e1fbb9a0dd/doc/logging.md, by:
```bash
export TSS2_LOG=fapi+NONE
```

Alternatively, you can compile the backend out of tpm2-pkcs11 via:
```bash
./configure --with-fapi=no
```

**Comment by Dvergatal** [comment]
Ok and what for FAPI is? Because i haven' found any relevant informations about it, just that it is some Feature API.

**Comment by williamcroberts** [comment]
> Ok and what for FAPI is? Because i haven' found any relevant informations about it, just that it is some Feature API.

Yes it's called the Feature API and its meant to be a higher level API to the TPM.

> 
> Ahh ok it is for storing keys in tpm instead in sqlite database. Correct me if i'm wrong.

Yes and no. It stores them to the file system instead, which conceptually isn't much different. The important thing, is those key blobs resident on disk are protected with both access controls as well as by the TPM itself.

**Comment by Dvergatal** [comment]
> > Ahh ok it is for storing keys in tpm instead in sqlite database. Correct me if i'm wrong.
> 
> Yes and no. It stores them to the file system instead, which conceptually isn't much different. The important thing, is those key blobs resident on disk are protected with both access controls as well as by the TPM itself.

Ok that is what i was not rally sure. Thx for the feedback.

**Comment by Dvergatal** [comment]
Ok, and the measured state is being made from what? Some board components?

**Comment by Dvergatal** [comment]
I have commented it out for test and the second problem is that ta/pkcs11/src/pkcs11_token.c throws error:
```bash
ta/pkcs11/src/pkcs11_token.c:7:10: fatal error: config.h: No such file or directory
    7 | #include <config.h>
```

**Comment by Dvergatal** [comment]
One more thing is it possible to delete already provisioned profile?

**Comment by williamcroberts** [comment]
> Ok, and the measured state is being made from what? Some board components?
> 

It's your boot firmware and OS measuring various bits into different PCRs, what goes into what I don't know off the top of my head. @idesai might know the details.

> BTW. I'm sorry for off-top, but i'm interested in some of tpm mechanism, because i have no idea on how to store securely pkcs11-tpm user password... On other pkcs#11 implementations i have secure boot implementation with some other mechanisms and i simply do not know.

Right now everything is pin based, but in the future we will have support for objects with policies that can be tied to PCR state. For now, you could perhaps do something like seal the pin to the tpm in an object that is protected by PCR state. Then unseal that object and use that pin to access the pkcs11 layer. The downside here, is that the pin is in the clear in the host OS, so if your attack model includes a compromised kernel or MITM attack, the attacker could then access pkcs11 layer without the measured state being correct.
